src/lib.rs

```diff
@@ -65,87 +66,44 @@ pub use crate::error::ValidationErrors;
 /// ```rust
 /// # fn main() -> Result<(), Box<dyn std::error::Error + 'static>> {
 /// # use serde_json::Value;
-/// # use jsonschema_valid::schemas;
-/// # let schema_json = "{}";
-/// # let your_json_data = "{}";
-/// let schema: Value = serde_json::from_str(schema_json)?;
-/// let data: Value = serde_json::from_str(your_json_data)?;
-///
-/// let validation = jsonschema_valid::validate(&data, &schema, Some(&schemas::Draft6), false);
-/// let errors = validation.get_errors();
-/// assert!(errors.is_empty());
-///
-/// # Ok(()) }
-/// ````
-pub fn validate(
-    instance: &Value,
-    schema: &Value,
-    draft: Option<&dyn schemas::Draft>,
-    validate_schema: bool,
-) -> error::ValidationErrors {
-    let mut errors = error::ValidationErrors::new();
-    config::Config::from_schema(schema, draft)
-        .unwrap()
-        .validate(instance, schema, &mut errors, validate_schema);
-    errors
-}
-
-/// Validates a given JSON instance against a given JSON schema, writing the
-/// errors to the given stream. draft may provide the schema draft to use. If
-/// not provided, it will be determined automatically from the schema.
-///
-/// # Arguments
-///
-/// * `stream`: An object to write errors to.
-/// * `instance`: The JSON document to validate
-/// * `schema`: The JSON schema to validate against
-/// * `draft`: The draft of the JSON schema specification to use. If `None`, the
-///   draft will be automatically determined from the `schema`.
-/// * `validate_schema`: When `true`, validate the schema against the metaschema
-///   first.
-///
-/// # Returns
-///
-/// * `Some(())`: No errors were found.
-/// * `None`: At least one error was found.
-///
-/// ## Example:
-///
-/// The following example validates some JSON data against a draft 6 JSON schema.
-///
-/// ```rust
-/// # fn main() -> Result<(), Box<dyn std::error::Error + 'static>> {
-/// # use serde_json::Value;
-/// # use jsonschema_valid::schemas;
-/// # use std::io;
+/// # use jsonschema_valid::{schemas, Config};
 /// # let schema_json = "{}";
 /// # let your_json_data = "{}";
 /// let schema: Value = serde_json::from_str(schema_json)?;
 /// let data: Value = serde_json::from_str(your_json_data)?;
+/// let cfg = jsonschema_valid::Config::from_schema(&schema, Some(&schemas::Draft6)).unwrap();
```

Comment:
This new way of doing things (returning a proper implementation of the `Iterator` Trait) that you can iterate over to get errors, feels a lot more like idiomatic Rust, and it greatly simplifies the code internally (and as you can see, reduces the number of lines of code, etc.)

However, this is the only way I've found to make the lifecycles work, which isn't a great public API.  In short, the problems is that the `cfg` object needs to live as long as the returned iterator from `jsonschema_valid::validate` (since the iterator needs to get more values, which needs access to the `cfg` object).  Any attempt to create that object *inside* of the validate function itself seems to elude me.  It runs into "self-referential struct" problems, since the iterator object needs to refer to an object (the `cfg` object) that belongs to itself.

Any thoughts? @badboy.  Generators seem like the perfect solution, but isn't in stable Rust yet.